panther_hardware_interfaces/src/panther_system_e_stop_manager.cpp

```diff
@@ -0,0 +1,167 @@
+        "controller is sending zero velocity commands or not sending any velocity commands before "
+        "attempting to reset the E-stop.");
+    }
+
+    try {
+      manager_resources_->gpio_controller->EStopReset();
+    } catch (const EStopResetInterrupted & e) {
+      throw std::runtime_error(
+        "The E-stop reset operation was halted because the E-stop was triggered again.");
+    } catch (const std::runtime_error & e) {
+      throw std::runtime_error(
+        "Error when trying to reset E-stop using GPIO: " + std::string(e.what()));
+    }
+
+    manager_resources_->roboteq_error_filter->SetClearErrorsFlag();
+  } else {
+    std::fprintf(stdout, "E-stop trigger operation is in progress, skipping reset.");
+  }
+
+  if (ConfirmEStopResetSuccessful()) {
+    e_stop_triggered_ = false;
+  }
+}
+
+bool EStopStrategyPTH10X::ReadEStopState()
+{
+  const bool motors_on =
+    manager_resources_->gpio_controller->IsPinActive(panther_gpiod::GPIOPin::STAGE2_INPUT);
+  const bool driver_error = manager_resources_->roboteq_error_filter->IsError();
+
+  if ((driver_error || !motors_on) && !e_stop_triggered_) {
+    TriggerEStop();
+  }
+
+  return e_stop_triggered_;
+}
+
+void EStopStrategyPTH10X::TriggerEStop()
+{
+  std::lock_guard<std::mutex> e_stop_lck(e_stop_manipulation_mtx_);
+  std::lock_guard<std::mutex> lck_g(*(manager_resources_->motor_controller_write_mtx));
+
+  try {
+    manager_resources_->motors_controller->TurnOnSafetyStop();
+  } catch (const std::runtime_error & e) {
+    throw std::runtime_error(
+      "Error when trying to set safety stop using CAN command: " + std::string(e.what()));
+  }
+
+  e_stop_triggered_ = true;
+}
+
+void EStopStrategyPTH10X::ResetEStop()
+{
+  if (e_stop_manipulation_mtx_.try_lock()) {
+    std::lock_guard<std::mutex> e_stop_lck(e_stop_manipulation_mtx_, std::adopt_lock);
+
+    if (!manager_resources_->motors_controller->AreVelocityCommandsNearZero()) {
+      throw std::runtime_error(
+        "Can't reset E-stop - last velocity commands are different than zero. Make sure that your "
+        "controller sends zero commands before trying to reset E-stop.");
+    }
+
+    if (!manager_resources_->gpio_controller->IsPinActive(panther_gpiod::GPIOPin::STAGE2_INPUT)) {
+      throw std::runtime_error("Can't reset E-stop - motors are not powered.");
+    }
+
+    if (!manager_resources_->roboteq_error_filter->IsError()) {
+      throw std::runtime_error("Can't reset E-stop - motor controller is in an error state.");
+    }
+
+    manager_resources_->roboteq_error_filter->SetClearErrorsFlag();
+  } else {
+    std::fprintf(stdout, "E-stop trigger operation is in progress, skipping reset.");
+  }
+
+  if (ConfirmEStopResetSuccessful()) {
+    e_stop_triggered_ = false;
+  }
+}
+
+void EStopManager::SetStrategy(std::unique_ptr<EStopStrategy> && strategy)
+{
+  strategy_ = std::move(strategy);
+  strategy_->SetManagerResources(resources_);
```

Comment:
I wonder if such a dependency injection is neccessary here.

pros:
- flexibility in strategy change
- no need to pass dependencies separately to dependencies constructor

cons:
- additional struct created and increased the code complexity
- the context is coupled with the strategies details (what if there is a new strategy that requires different attributes)

I'm not saying this approach is wrong, it's more like a discussion point.
Did you find a similar implementation as an example?